rust/signed_doc/src/validator/rules/content_type.rs

```diff
@@ -48,13 +48,93 @@ impl ContentTypeRule {
 
         Ok(true)
     }
+
+    /// Validates the provided `content` bytes to be a defined `ContentType`.
+    pub(crate) fn validate(&self, content: &[u8]) -> anyhow::Result<()> {
+        match self.exp {
+            ContentType::Json => {
+                if let Err(e) = serde_json::from_slice::<serde_json::Value>(content) {
+                    anyhow::bail!("Invalid {} content: {e}", self.exp)
+                }
+            },
+            ContentType::Cbor => {
+                let mut decoder = minicbor::Decoder::new(content);
+
+                decoder.skip()?;
+
+                if decoder.position() != content.len() {
+                    anyhow::bail!("Unused bytes remain in the input after decoding")
+                }
+            },
+        }
+        Ok(())
+    }
 }
 
 #[cfg(test)]
 mod tests {
+    use proptest::prelude::*;
+
     use super::*;
     use crate::Builder;
 
+    #[test]
+    fn content_type_validate_test() {
+        let json_rule = ContentTypeRule {
+            exp: ContentType::Json,
+        };
+        let cbor_rule = ContentTypeRule {
+            exp: ContentType::Cbor,
+        };
+
+        let json_bytes = serde_json::to_vec(&serde_json::Value::Null).unwrap();
+        assert!(json_rule.validate(&json_bytes).is_ok());
+        assert!(cbor_rule.validate(&json_bytes).is_err());
+
+        let cbor_bytes = minicbor::to_vec(minicbor::data::Token::Null).unwrap();
+        assert!(json_rule.validate(&cbor_bytes).is_err());
+        assert!(cbor_rule.validate(&cbor_bytes).is_ok());
+    }
+
+    #[test]
+    fn cbor_with_trailing_bytes_test() {
+        // valid cbor: {1: 2} but with trailing 0xff
+        let mut buf = Vec::new();
+        let mut enc = minicbor::Encoder::new(&mut buf);
+        enc.map(1).unwrap().u8(1).unwrap().u8(2).unwrap();
+        buf.push(0xFF); // extra byte
+
+        let cbor_rule = ContentTypeRule {
+            exp: ContentType::Cbor,
+        };
+
+        assert!(cbor_rule.validate(&buf).is_err());
+    }
+
+    #[test]
+    fn malformed_cbor_bytes_test() {
+        // 0xa2 means a map with 2 key-value pairs, but we only give 1 key
+        let invalid_bytes = &[0xA2, 0x01];
+
+        let cbor_rule = ContentTypeRule {
+            exp: ContentType::Cbor,
+        };
+
+        assert!(cbor_rule.validate(invalid_bytes).is_err());
+    }
+
+    proptest! {
+        #[test]
+        fn invalid_cbor_from_random_data_test(ref data in proptest::collection::vec(any::<u8>(), 1..100)) {
```

Comment:
Whats the proof that some random data isn't going to actually be valid cbor?
If we don't know the data is not valid cbor, then the test may produce random bytes which are valid cbor.